src/pydvl/utils/parallel/futures/ray.py

```diff
@@ -0,0 +1,324 @@
+                        self.work_queue.put_nowait(work_item)
+                        return
+                    logger.debug("moving work item from pending queue to work queue")
+                    self.work_queue.put(work_item)
+                    del work_item
+
+    def submit_work_item(self) -> None:
+        with self.queue_lock:
+            logger.debug("work item manager thread acquired queue lock")
+            # Try to get a work item
+            # If it is None, we break from the loop.
+            # Otherwise, we submit the future.
+            try:
+                logger.debug("getting work item from work queue")
+                work_item = self.work_queue.get_nowait()
+                if work_item is None:
+                    return
+            except queue.Empty:
+                logger.debug("work queue is empty")
+                return
+        logger.debug("Submitting work item")
+        work_item.run()
+        # Delete references to object
+        del work_item
+
+    def is_shutting_down(self) -> bool:
+        # Check whether we should start shutting down the executor.
+        executor = self.executor_reference()
+        # No more work items can be added if:
+        #   - The executor that owns this worker has been collected OR
+        #   - The executor that owns this worker has been shutdown.
+        with self.shutdown_lock:
+            logger.debug("work item manager thread acquired shutdown lock")
+            if executor is None or executor._shutdown:
+                # Flag the executor as shutting down as early as possible if it
+                # is not gc-ed yet.
+                if executor is not None:
+                    executor._shutdown = True
+                return True
+            return False
+
+    def flag_executor_shutting_down(self):
+        # Flag the executor as shutting down and cancel remaining tasks if
+        # requested as early as possible if it is not gc-ed yet.
+        executor = self.executor_reference()
+        with self.shutdown_lock:
+            logger.debug("work item manager thread acquired shutdown lock")
+            if executor is not None:
+                executor._shutdown = True
+                # Cancel pending work items if requested.
+                if executor._cancel_pending_futures:
+                    logger.debug("forcefully cancelling futures")
+                    # Drain all work items from the queues,
+                    # and then cancel their associated futures.
+                    # We empty the pending queue first.
+                    while True:
+                        with self.queue_lock:
+                            try:
+                                work_item = self.pending_queue.get_nowait()
+                            except queue.Empty:
+                                break
+                            if work_item is not None:
+                                work_item.future.cancel()
```

Comment:
As mentioned in https://github.com/appliedAI-Initiative/pyDVL/pull/329#discussion_r1141327845, this is probably a noop for ray's returned futures. One must either wrap those to make `cancel()` work, or call `ray.cancel()` on them.

Reply:
I changed things a bit in the code. The work items that are still in the queues were not yet submitted to the ray cluster so there is no need for that.

Perhaps I should, like previously, keep the submitted futures in a weakset and call `ray.cancel()` on them.
I will add that.

Reply:
Done.